samples/n5010-ctl/n5010-ctl.c

```diff
@@ -0,0 +1,504 @@
+
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <errno.h>
+#include <error.h>
+#include <time.h>
+#include <signal.h>
+#include <getopt.h>
+#include <opae/fpga.h>
+#include <uuid/uuid.h>
+#include <endian.h>
+#include "argsfilter.h"
+
+#define DFH_EOL(h) ((h >> 40) & 1)
+#define DFH_NEXT(h) ((h >> 16) & 0xffffff)
+#define DFH_TYPE(h) ((h >> 60) & 0xf)
+#define DFH_TYPE_AFU 1
+
+#define MAX_PORT 3
+#define INVALID_PORT (MAX_PORT + 1)
+#define CVL0_QSFP01_SWITCH 0x28
+#define CVL1_QSFP23_SWITCH 0x30
+
+char const *default_guid = "c48d4e2a-9121-497d-adc7-3640729ec6f7";
+
+// Define the function to be called when ctrl-c (SIGINT) signal is sent to
+// process
+static volatile bool stop = false;
+void signal_callback_handler(int signum)
+{
+	switch (signum) {
+	case SIGINT:
+		stop = true;
+		break;
+	default:
+		break;
+	}
+}
+
+struct n5010;
+
+struct n5010_test {
+	const char *name;
+	fpga_result (*func)(struct n5010 *n5010);
+};
+
+struct n5010 {
+	fpga_properties filter;
+	fpga_token token;
+	fpga_handle handle;
+	fpga_guid guid;
+	uint64_t base;
+	const struct n5010_test *test;
+	uint64_t port;
+	bool debug;
+};
+
+static fpga_result fpga_setread_switch(struct n5010 *n5010);
+static fpga_result fpga_read_switch(struct n5010 *n5010);
+static fpga_result fpga_run(struct n5010 *n5010);
+
+static const struct n5010_test n5010_test[] = {
+	{
+		.name = "setsw",
+		.func = fpga_setread_switch,
+	},
+	{
+		.name = "readsw",
+		.func = fpga_read_switch,
+	},
+	{
+		.name = "runsw",
+		.func = fpga_run,
+	},
+};
+
+
+static fpga_result fpga_open(struct n5010 *n5010)
+{
+	fpga_result res;
+	uint32_t num = 0;
+	res = fpgaPropertiesSetObjectType(n5010->filter, FPGA_ACCELERATOR);
+	if (res != FPGA_OK) {
+		fprintf(stderr, "failed to set object type: %s\n",
+	fpgaErrStr(res)); goto error;
+	}
+	res = fpgaPropertiesSetGUID(n5010->filter, n5010->guid);
+	if (res != FPGA_OK) {
+		fprintf(stderr, "failed to set guid: %s\n", fpgaErrStr(res));
+		goto error;
+	}
+
+	res = fpgaEnumerate(&n5010->filter, 1, &n5010->token, 1, &num);
+	if (res != FPGA_OK) {
+		fprintf(stderr, "failed to enumerate fpga: %s\n",
+			fpgaErrStr(res));
+		goto error;
+	}
+
+	if (num == 0) {
+		res = FPGA_NOT_FOUND;
+		fprintf(stderr, "failed to find fpga with guid: %s\n",
+			fpgaErrStr(res));
+		goto error;
+	}
+	{
+		fpga_properties prop;
+		uint8_t bus;
+		fpgaGetProperties(n5010->token, &prop);
+		fpgaPropertiesGetBus(prop, &bus);
+		fpgaDestroyProperties(&prop);
+	}
+
+	res = fpgaOpen(n5010->token, &n5010->handle, FPGA_OPEN_SHARED);
+
+	if (res != FPGA_OK) {
+		fprintf(stderr, "failed to open fpga: %s\n", fpgaErrStr(res));
+		goto error;
+	}
+
+	res = fpgaMapMMIO(n5010->handle, 0, NULL);
+	if (res != FPGA_OK) {
+		fprintf(stderr, "failed to map io memory: %s\n",
+			fpgaErrStr(res));
+		goto error;
+	}
+
+error:
+	return res;
+}
+
+static void fpga_close(struct n5010 *n5010)
+{
+	fpga_result res;
+
+	if (n5010->handle != NULL) {
```

Comment:
please merge if unmapMMIO and fpgaClose into  one if statement 
if(n5010->handel) {  '
fpgaunmoMMIO() 
fpgaclose()
}

Reply:
Done